### ISSUE TITLE
Microsoft.Common.props\ImportBefore\* is evaluated before setting $(VisualStudioVersion)

### ISSUE THREAD

**Issue by iskiselev** [Microsoft/msbuild#1678]
I'm trying to compile (rebuild) solution, that contains several project with nuget dependencies in project.json (with old, non-sdk based .csproj) using msbuild from VS2017 RC4+26206.0. (I call msbuild using command line "C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\MSBuild\15.0\Bin\MSBuild.exe").
Previously I successfully build project from Visual Studio 2017, so all nuget dependecies should already be restored.
I've configured project to Treat all warnings as error, and output documentation. Also, there is some public classes without xml docs, so there will be CS1591 error by default.
To mute this error, I've attached nuget dependecy (through package.json) with build\DepName.props with content:
```
<?xml version="1.0" encoding="utf-8"?>
<Project ToolsVersion="4.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
    <PropertyGroup>
        <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
        <WarningsNotAsErrors>$(WarningsNotAsErrors);0618;0612</WarningsNotAsErrors>
	<NoWarn>$(NoWarn);1591</NoWarn>
    </PropertyGroup>
</Project>
```

Visual Studio compile this without any errors, as expected. But with command line msbuild build it ends up with error CS1591.

If you need, I may provide reproducing example.

**Comment by rainersigwald** [comment]
A repro example would help; if you can't provide that can you share a diagnostic level log both from within VS and from the command line?

**Comment by iskiselev** [comment]
Please take repro from https://github.com/iskiselev/msbuild_projecJson
This projects build OK from VS 2017, but fails using command line:
```
"C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\MSBuild\15.0\Bin\MSBuild.exe" /t:Rebuild /p:Configuration=Debug ProjectJsonBuild.sln
```

**Comment by rainersigwald** [comment]
I don't reproduce the failure. After building in VS, I get success:

```
s:\work\iskiselev_missed_target>msbuild /t:Rebuild /p:Configuration=Debug ProjectJsonBuild.sln /flp:v=diag
Microsoft (R) Build Engine version 15.1.545.13942
Copyright (C) Microsoft Corporation. All rights reserved.

Building the projects in this solution one at a time. To enable parallel build, please add the "/m" switch.
Build started 2/7/2017 5:43:59 PM.
Project "s:\work\iskiselev_missed_target\ProjectJsonBuild.sln" on node 1 (Rebuild target(s)).
ValidateSolutionConfiguration:
  Building solution configuration "Debug|Any CPU".
Project "s:\work\iskiselev_missed_target\ProjectJsonBuild.sln" (1) is building "s:\work\iskiselev_missed_target\Project
JsonBuild\ProjectJsonBuild.csproj" (2) on node 1 (Rebuild target(s)).
CoreClean:
  Deleting file "s:\work\iskiselev_missed_target\ProjectJsonBuild\bin\Debug\ProjectJsonBuild.xml".
  Deleting file "s:\work\iskiselev_missed_target\ProjectJsonBuild\bin\Debug\ProjectJsonBuild.exe.config".
  Deleting file "s:\work\iskiselev_missed_target\ProjectJsonBuild\bin\Debug\ProjectJsonBuild.exe".
  Deleting file "s:\work\iskiselev_missed_target\ProjectJsonBuild\bin\Debug\ProjectJsonBuild.pdb".
  Deleting file "s:\work\iskiselev_missed_target\ProjectJsonBuild\obj\Debug\ProjectJsonBuild.csprojResolveAssemblyRefer
  ence.cache".
  Deleting file "s:\work\iskiselev_missed_target\ProjectJsonBuild\obj\Debug\ProjectJsonBuild.exe".
  Deleting file "s:\work\iskiselev_missed_target\ProjectJsonBuild\obj\Debug\ProjectJsonBuild.pdb".
GenerateBindingRedirects:
  No suggested binding redirects from ResolveAssemblyReferences.
GenerateTargetFrameworkMonikerAttribute:
Skipping target "GenerateTargetFrameworkMonikerAttribute" because all output files are up-to-date with respect to the i
nput files.
CoreCompile:
  C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\MSBuild\15.0\bin\Roslyn\csc.exe /noconfig /nowarn:1591
  ,1701,1702 /nostdlib+ /platform:anycpu32bitpreferred /errorreport:prompt /warn:4 /doc:bin\Debug\ProjectJsonBuild.xml
  /define:DEBUG;TRACE /highentropyva+ /reference:"C:\Program Files (x86)\Reference Assemblies\Microsoft\Framework\.NETF
  ramework\v4.5.2\Microsoft.CSharp.dll" /reference:"C:\Program Files (x86)\Reference Assemblies\Microsoft\Framework\.NE
  TFramework\v4.5.2\mscorlib.dll" /reference:"C:\Program Files (x86)\Reference Assemblies\Microsoft\Framework\.NETFrame
  work\v4.5.2\System.Core.dll" /reference:"C:\Program Files (x86)\Reference Assemblies\Microsoft\Framework\.NETFramewor
  k\v4.5.2\System.Data.DataSetExtensions.dll" /reference:"C:\Program Files (x86)\Reference Assemblies\Microsoft\Framewo
  rk\.NETFramework\v4.5.2\System.Data.dll" /reference:"C:\Program Files (x86)\Reference Assemblies\Microsoft\Framework\
  .NETFramework\v4.5.2\System.dll" /reference:"C:\Program Files (x86)\Reference Assemblies\Microsoft\Framework\.NETFram
  ework\v4.5.2\System.Net.Http.dll" /reference:"C:\Program Files (x86)\Reference Assemblies\Microsoft\Framework\.NETFra
  mework\v4.5.2\System.Xml.dll" /reference:"C:\Program Files (x86)\Reference Assemblies\Microsoft\Framework\.NETFramewo
  rk\v4.5.2\System.Xml.Linq.dll" /debug+ /debug:full /filealign:512 /optimize- /out:obj\Debug\ProjectJsonBuild.exe /rul
  eset:"C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\Team Tools\Static Analysis Tools\\Rule Sets\Mini
  mumRecommendedRules.ruleset" /subsystemversion:6.00 /target:exe /warnaserror+ /utf8output Program.cs Properties\Assem
  blyInfo.cs "C:\Users\raines\AppData\Local\Temp\.NETFramework,Version=v4.5.2.AssemblyAttributes.cs" /warnaserror-:0618
  ,0612
  Using shared compilation with compiler from directory: C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise
  \MSBuild\15.0\bin\Roslyn
_CopyAppConfigFile:
  Copying file from "App.config" to "bin\Debug\ProjectJsonBuild.exe.config".
CopyFilesToOutputDirectory:
  Copying file from "obj\Debug\ProjectJsonBuild.exe" to "bin\Debug\ProjectJsonBuild.exe".
  ProjectJsonBuild -> s:\work\iskiselev_missed_target\ProjectJsonBuild\bin\Debug\ProjectJsonBuild.exe
  Copying file from "obj\Debug\ProjectJsonBuild.pdb" to "bin\Debug\ProjectJsonBuild.pdb".
Done Building Project "s:\work\iskiselev_missed_target\ProjectJsonBuild\ProjectJsonBuild.csproj" (Rebuild target(s)).

Done Building Project "s:\work\iskiselev_missed_target\ProjectJsonBuild.sln" (Rebuild target(s)).


Build succeeded.
    0 Warning(s)
    0 Error(s)

Time Elapsed 00:00:02.25
```

**Comment by iskiselev** [comment]
I've added my diagnostic msbuild log, if it will help (https://github.com/iskiselev/msbuild_projecJson/blob/master/msbuild.log). I'll try to test on another machine also. Thank you for your help.

**Comment by iskiselev** [comment]
Some additional notes: it works correct, when I run msbuild from Developer Command Propt for VS 2017 RC.
It doesn't work if I run it using full path to msbuild from cmd. It worked with msbuild v.14 even from simple command prompt. Could you advice which environment parameters may matter?

**Comment by rainersigwald** [comment]
Ah, that explains the difference! #1461 made some changes in this space, and I would have expected it to work. It _almost_ does, but by comparing the output of the MSBuild preprocessor between a plain command prompt and a "Developer Command Prompt for VS 2017 RC", I can see that the NuGet targets don't get imported. That's because `$(VisualStudioVersion)` isn't set from the environment.

Everything else works, because a following line in common.props

```xml
<Import Project="$(MSBuildExtensionsPath)\$(MSBuildToolsVersion)\Microsoft.VisualStudioVersion.v*.Common.props" Condition="'$(VisualStudioVersion)' == ''">
```

imports a file that sets `VisualStudioVersion` for the rest of the evaluation. But extension points like NuGet that are imported via `$(MSBuildExtensionsPath)\$(MSBuildToolsVersion)\Imports\Microsoft.Common.props\ImportBefore\*` are before that point.

I think it would be reasonable to move the `ImportBefore` machinery after the version-setting machinery, but that has compat impact and I don't want to do so for 15.1.

I'll update the title of this bug to represent that more general case.

@iskiselev You should be able to work around this limitation by setting the environment variable `VisualStudioVersion` or passing `/property:VisualStudioVersion=15.0` on the command line.

**Comment by rainersigwald** [comment]
Fixed by #2258.
